docs/rfcs/0007-export-direction.md

```diff
@@ -0,0 +1,221 @@
+
+Hence, nether option can be considered less verbose in the general case.
+
+### Ease of reading
+
+Ideally, one should be able to read a River config file from top to bottom easily. In addition:
+* It should be easy to see where a component's inputs are coming from.
+* It should be easy to see where a component's outputs are going to.
+
+It is not possible to achieve both in practice, just by reading a config file. That's because we only specify either where data came from, or where it's going to. It would be redundant to specify this in both the source component and the destination component. The only way to achieve both of those things is to use a UI tool for visualizing the graph.
+
+Approach 1 makes tracing dependencies easy for some components like this one...
+```
+prometheus.scrape "default" {
+  targets    = discovery.kubernetes.pods.targets
+  forward_to = [prometheus.remote_write.default.receiver]
+}
+```
+
+... but it makes it extra hard for other components such as this:
+```
+discovery.kubernetes "pods" {
+  role = "pod"
+}
+```
+
+It may have been easier to read the file if `discovery.kubernetes` listed it outputs like so:
+```
+discovery.kubernetes "pods" {
+  role = "pod"
+  targets = [prometheus.scrape.default]
+}
+```
+
+It could also help readability if we advise users to put downstream components at the top, or at the bottom, depending on whether we go with approach 2 or 3.
+
+### Could any approach lead to composition issues?
+
+In River, it is possible to set any of a component's attributes via exports from another component. If the exports of component A are actually used to feed inputs to component A, then component B would not be able to use all exports from component A to feed a component B.
+
+It would be much simpler if we could just say to the users that "every argument can be set via another component's exports - provided that the types match".
+
+Approach 1 can make it harder for users to figure out how to compose components, which then means we have to provide examples for lots of potential use cases. It would be better if there is no need for excessive use of examples.
+
+### Learning curve
+
+The more consistent our approach, the easiest it is to explain in documentation. In order to link two components, at the moment sometimes users specify the next component in the "outputs" section, and sometimes they specify the previous component in the "inputs" section. This can be confusing. 
+
+### Using components from a River file which cannot or should not be modified (River libraries?)
+
+In this case, it is best for the upstream component not to specify all the downstream components which use its output. That way components which need data that the upstream exports could simply use it in any of the downstream's attributes.
+
+### Preventing dangling references to shut down components
```

Comment:
@rfratto mentioned in a global comment:
> I’m concerned we’re not touching enough on how it will be possible to reverse the direction of some components in a way which is easy to maintain and that isn’t prone to bugs.

I'll continue the discussion in this thread because it might get lengthy :) Before we discuss a solution, I'd like to firstly discuss the problem to make sure we understand it.

### What bugs could happen if we change the exports direction?
I suppose that the bugs @rfratto is referring to are mostly dangling references? Please let me know if there could be other ones too? Below I'll include my thoughts on dangling references.

### Can we decouple the representation of the graph in config from how the graph is built in the code?
I am surprised that our ability to treat dangling references is related to the way the graph is described in config. I would have expected this to be decoupled more. Do we want to try decoupling it so that we can do future changes with less friction?

Regardless of how the graph is represented in config, IMO all users infer from it is that there is an edge between two nodes on a graph. We could handle the internals as we see fit. The config representation is mostly for readability and aesthetics.

### When could we end up with dangling references? 
I can think of two situations:

1. A misconfiguration during a config reload. This is what this issue was patching:
#2710
Maybe such reload problems could be fixed for all components via:
#2709
2. A component stops working spontaneously. What should we do then?
    * Should we restart the component?
    * Restart the agent process? We almost certainly can't do that, because if we shut down we don't know if the Agent was set up with some sort of restart policy.
    * Is there any other way to attempt repairing it?
    * What do the components which depend on the failing component do? Do they restart? It could be that they are the real cause of the problem and maybe they need to be restarted?
    * What about buildup of data to the failing component?

### Have a solution at least as good as the one today. What issues does the current implementation prevent?
The current implementation prevents some dangling references. E.g. if `prometheus.remote_write` shuts down, it will remove references to other components. However, other components such as `prometheus.scrape` would still have a dangling reference to it. And as shown by ##2710, the current implementation could lead to bugs as well. Maybe we can't find a bug-free solution, but could make something at least as good as the current one.

### What should the user experience look like?
When we agree on the desired user experience during component failure, I think we could start discussing potential solutions.

Reply:
> I suppose that the bugs @rfratto is referring to are mostly dangling references? 

Let me give an example of what I'm talking about here:

Today, components which expose current information are pull-based: 

```river 
discovery.kubernetes "pods" {
  role = "pod" 
}

prometheus.scrape "pods" {
  // Pull from discovery.kubernetes.pods.
  targets = discovery.kubernetes.pods.targets 
}
```

If `discovery.kubernetes.pods` is removed from the configuration file, it will cause a evaluation error with prometheus.scrape.pods, and prometheus.scrape.pods will be left in its last valid state until the evaluation error is resolved. Once the reference is removed from prometheus.scrape.pods, everything becomes healthy again. 

Let's assume now that the direction is reversed:

```river 
discovery.kubernetes "pods" {
  role = "pod" 

  // Push targets to prometheus.scrape.pods. 
  send_to = [prometheus.scrape.pods.targets] 
}

prometheus.scrape "pods" { }
```

The implementation of this requires: 

1. Some kind of stateful operation for discovery.kubernetes.pods to notify prometheus.scrape.pods of new targets. 
2. prometheus.scrape.pods must be able to correctly combine targets received from multiple senders, implying it must cache targets. 

The implementation means that push-based discovery component authors must find a way to unregister the latest targets from whoever they sent it to. If they don't, and discovery.kubernetes.pods is removed from the config file, the targets leak, and prometheus.scrape.pods holds on to targets forever with no signal to the user; the user doesn't know things are unhealthy. 

This isn't an issue with pushing events, but it's a complexity introduced when you're talking about synchronizing state via pushes. 

Does that make sense?

Reply:
The same problem exists with swapping event-based components to be pull-based instead:

```river 
prometheus.scrape "example" {
  ... 
}

prometheus.remote_write "example" {
  read_from = [prometheus.scrape.example.output]
}
```

This implies that:

1. prometheus.remote_write.example must subscribe to prometheus.scrape.example. 
2. prometheus.scrape.example must keep track of all components which is subscribed. 
3. prometheus.remote_write.example must unsubscribe on shutdown, otherwise prometheus.scrape.example will keep a dangling reference forever.

Reply:
So, to summarize, the issue is that swapping the direction of these components _is_ decoupled from the controller; the current implementation being coupled with the controller means that components can't be implemented in a way which causes memory leaks in _other components._

Reply:
So when I say 

> I’m concerned we’re not touching enough on how it will be possible to reverse the direction of some components in a way which is easy to maintain and that isn’t prone to bugs.

I mean we need to find a way to implement an event system (for everything-is-push) or a subscription system (for everything-is-pull) which means that it's either impossible or extremely difficult for component authors to cause a memory leak in an unrelated component.

Reply:
I'm slightly confused - on one hand we must do pull and cannot do push for targets, but on the other hand we must to push and we cannot do pull for events. What is the essential difference between targets and events that makes it so? 

I get it that particularly for one-to-many connections, the push model is easier to manage. But I don't get it why push model cannot be used with targets - can't we treat a discovery of targets like an event that is sent to many components on `forward_to` list? If the implementation can cause a leak for targets, why are there no leaks for 'events'?

Reply:
@thampiotr The primary difference is that events are stateless, while targets are stateful, as they declare the current state of things. 

> But I don't get it why push model cannot be used with targets - can't we treat a discovery of targets like an event that is sent to many components on forward_to list? 

I've tried to describe the challenges with it here: 

https://github.com/grafana/agent/pull/3325#discussion_r1200459648

Was there any part of this that wasn't clear that I can help clarify? 

> If the implementation can cause a leak for targets, why are there no leaks for 'events'?

Because events aren't stateful; you don't have to persist all of the events you've received over time to build an aggregated combined list like you have to with targets.

Reply:
That's a fair point, but if you were able to hide this behind the scenes for pull-based approach, it should be possible to hide it also for a push-based approach. Right now `concat` will combine most recently exported values and call `Update` for components that consume it. We could have a receiver that does the opposite of fan-out - return a new channel for every producer that wants to write to it and combine the most recently sent values for the consumer (while timing out and defaulting to 0 for dead channels).

What is kinda interesting is that we already have a de-facto push-based model for targets, because River pushes the updates with a call to `Update` function ;)